dns_server/src/crawler/mod.rs

```diff
@@ -0,0 +1,560 @@
+                tx.add_address(&address.to_string()).expect("adding address must succeed");
+                tx.commit().expect("tx commit must succeed");
+
+                if let Some(ip) = dns_ip {
+                    command_tx.send(ServerCommands::AddAddress(ip)).expect("sending must succeed");
+                }
+
+                address_data.fail_count = 0;
+                address_data.was_reachable = true;
+            }
+            AddressState::Disconnected => {
+                address_data.fail_count += 1;
+            }
+        }
+    }
+
+    /// Returns true when it is time to attempt a new outbound connection
+    fn connect_now(now: tokio::time::Instant, address_data: &AddressData) -> bool {
+        match address_data.state {
+            AddressState::Connected | AddressState::Connecting => false,
+            AddressState::Disconnected if address_data.was_reachable => {
+                let age = now.duration_since(address_data.state_updated_at);
+
+                match address_data.fail_count {
+                    0 => true,
+                    1 => age > Duration::from_secs(60),
+                    2 => age > Duration::from_secs(360),
+                    3 => age > Duration::from_secs(3600),
+                    4 => age > Duration::from_secs(3 * 3600),
+                    5 => age > Duration::from_secs(6 * 3600),
+                    6 => age > Duration::from_secs(12 * 3600),
+                    _ => age > Duration::from_secs(24 * 3600),
+                }
+            }
+            AddressState::Disconnected => {
+                // The address was never reachable, try to connect just once
+                address_data.fail_count == 0
+            }
+        }
+    }
+
+    /// Returns true if the address should be kept in memory
+    fn retain_address(
+        now: tokio::time::Instant,
+        address: &N::Address,
+        address_data: &mut AddressData,
+        storage: &mut S,
+    ) -> bool {
+        if address_data.state == AddressState::Disconnected
+            && address_data.was_reachable
+            && address_data.fail_count >= PURGE_REACHABLE_FAIL_COUNT
+        {
+            log::debug!("purge old (once reachable) address {}", address.to_string());
+
+            let mut tx = storage.transaction_rw().expect("tx must succeed");
+            tx.del_address(&address.to_string()).expect("adding address must succeed");
+            tx.commit().expect("tx commit must succeed");
+
+            return false;
+        }
+
+        if address_data.state == AddressState::Disconnected
+            && !address_data.was_reachable
+            && address_data.fail_count > 0
+            && now.duration_since(address_data.state_updated_at) >= PURGE_UNREACHABLE_TIME
+        {
+            log::debug!("purge old (unreachable) address {}", address.to_string());
+
+            return false;
+        }
+
+        true
+    }
+
+    /// Address list maintenance
```

Comment:
Please type more description to what's supposed to happen in this function.

Reply:
Sure, updated.